From 1d4de038a59b329ecf3dba851612792258d29cec Mon Sep 17 00:00:00 2001
From: Carsten Fuchs <carsten.fuchs@cafu.de>
Date: Sat, 16 Aug 2014 14:54:17 +0200
Subject: [PATCH 10/10] Change wxDataViewMainWindow::Cleared() to keep the
 existing internal tree if possible.

Destroying and re-allocating the entire tree was unnecessary and especially it
forewent the information which nodes were open/expanded, causing the entire
tree to be all-closed again.
---
 src/generic/datavgen.cpp | 52 ++++++++++++++++++------------------------------
 1 file changed, 19 insertions(+), 33 deletions(-)

diff --git a/src/generic/datavgen.cpp b/src/generic/datavgen.cpp
index 18b7bc8..110ba98 100644
--- a/src/generic/datavgen.cpp
+++ b/src/generic/datavgen.cpp
@@ -766,7 +766,6 @@ public:
     // wxDataViewTreeNode * GetTreeNodeByItem( const wxDataViewItem & item );
 
     // Methods for building the mapping tree
-    void BuildTree( wxDataViewModel  * model );
     void DestroyTree();
     void HitTest( const wxPoint & point, wxDataViewItem & item, wxDataViewColumn* &column );
     wxRect GetItemRect( const wxDataViewItem & item, const wxDataViewColumn* column );
@@ -2528,20 +2527,31 @@ bool wxDataViewMainWindow::ValueChanged( const wxDataViewItem & item, unsigned i
     return true;
 }
 
-bool wxDataViewMainWindow::Cleared()
+bool wxDataViewMainWindow::Cleared()    // "(re-)sync to user model" (or "Update") would be a better name...
 {
-    DestroyTree();
+    if (!GetModel() || GetModel()->IsVirtualListModel())
+    {
+        wxDELETE(m_root);
+        m_root = NULL;
+    }
+
     m_selection.Clear();
     m_currentRow = (unsigned)-1;
+    m_count = 0;
 
     if (GetModel())
     {
         SortPrepare();
-        BuildTree( GetModel() );
-    }
-    else
-    {
-        m_count = 0;
+
+        if (!GetModel()->IsVirtualListModel())
+        {
+            if (!m_root)
+                m_root = wxDataViewTreeNode::CreateRootNode();
+
+            m_root->AlignTree(GetModel());
+        }
+
+        m_count = -1;
     }
 
     GetOwner()->InvalidateColBestWidths();
@@ -3431,23 +3441,6 @@ int wxDataViewMainWindow::GetRowByItem(const wxDataViewItem & item) const
     }
 }
 
-void wxDataViewMainWindow::BuildTree(wxDataViewModel * model)
-{
-    DestroyTree();
-
-    if (GetModel()->IsVirtualListModel())
-    {
-        m_count = -1;
-        return;
-    }
-
-    m_root = wxDataViewTreeNode::CreateRootNode();
-
-    SortPrepare();
-    m_root->AlignTree(model);
-    m_count = -1;
-}
-
 void wxDataViewMainWindow::DestroyTree()
 {
     if (!IsVirtualList())
@@ -4561,14 +4554,7 @@ bool wxDataViewCtrl::AssociateModel( wxDataViewModel *model )
         m_notifier = NULL;
     }
 
-    m_clientArea->DestroyTree();
-
-    if (model)
-    {
-        m_clientArea->BuildTree(model);
-    }
-
-    m_clientArea->UpdateDisplay();
+    m_clientArea->Cleared();
 
     return true;
 }
-- 
1.8.4.msysgit.0

